### PR TITLE
Proposal to add `AudioFileSourceFunction` which can generate sound from functions

### PR DESCRIPTION
**Pull request by hideakitai**
Hi @earlephilhower , thank you for this great audio library. I've made some extension named `AudioFileSourceFunction`  that can easily play "raw" audio like sine wave which is generated by functions which has 1 float argument(time) and 1 float return(amplitude). 

```C++
  // AudioFileSourceFunction can generate sound from function
  file = new AudioFileSourceFunction(8.); // [sec]
  file->addAudioGenerators([](const float time) {
    return sin(TWO_PI * 440.f * time);  // C
  });
  // same as other examples
  out = new AudioOutputI2SNoDAC();
  wav = new AudioGeneratorWAV();
  wav->begin(file, out)
```

`AudioFileSourceFunction` generates the WAV file header and its audio data will be generated each time based on the registered function when requested.

```C++
  file = new AudioFileSourceFunction(sec, channels, hz, bit_per_sample);
  // channels   : default = 1
  // hz         : default = 8000 (8000, 11025, 22050, 44100, 48000, etc.)
  // bit/sample : default = 16 (8, 16, 32)
```

The [example for AudioFileSourceFunction](https://github.com/hideakitai/ESP8266Audio/blob/add_source_function/examples/PlayWAVFromFunction/PlayWAVFromFunction.ino) is also added. I'm glad if you see whether this extension can works in your great library.

Thanks
hideakitai

### CODE REVIEW

File: src/AudioFileSourceFunction.cpp [hunk 1]
```diff
@@ -0,0 +1,128 @@
+#include "AudioFileSourceFunction.h"
```
Comment (earlephilhower) [inline src/AudioFileSourceFunction.cpp:1]: Can you please add the GPL header with your name and (c)?

File: src/AudioFileSourceFunction.h [hunk 2]
```diff
@@ -0,0 +1,99 @@
+#ifndef _AUDIOFILESOURCEFUNCTION_H
```
Comment (earlephilhower) [inline src/AudioFileSourceFunction.h:1]: GPL header, please

**Review by earlephilhower — CHANGES_REQUESTED** [review]
Very ingenious use of lambdas!  Add the appropriate headers and I'd be happy to merge.

This looks like it would also be useful to make things like DTMF tones as some people have requested.

**Comment by hideakitai** [comment]
Thanks for reviewing. I've added the GPL headers and also some supplements to example for someone who is not familiar to lambdas.

- Use lambda capture by default
- Add usage of pre-defined function instead of lambda

If you have any feedback, please let me know. Thank you!

**Review by earlephilhower — APPROVED** [review]

**Pull request merged**